src/network/networkd-sysctl.c

```diff
@@ -5,17 +5,193 @@
 #include <linux/if_arp.h>
 
 #include "af-list.h"
+#include "cgroup-util.h"
+#include "fd-util.h"
 #include "missing_network.h"
 #include "networkd-link.h"
 #include "networkd-lldp-tx.h"
 #include "networkd-manager.h"
 #include "networkd-ndisc.h"
 #include "networkd-network.h"
 #include "networkd-sysctl.h"
+#include "path-util.h"
+#include "sd-messages.h"
 #include "socket-util.h"
 #include "string-table.h"
 #include "sysctl-util.h"
 
+#if HAVE_VMLINUX_H
+
+#include "bpf-link.h"
+
+#include "bpf/sysctl_monitor/sysctl-monitor-skel.h"
+#include "bpf/sysctl_monitor/sysctl-write-event.h"
+
+static struct sysctl_monitor_bpf *sysctl_monitor_bpf_free(struct sysctl_monitor_bpf *obj) {
+        sysctl_monitor_bpf__destroy(obj);
+        return NULL;
+}
+
+static struct ring_buffer *rb_free(struct ring_buffer *rb) {
+        sym_ring_buffer__free(rb);
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct sysctl_monitor_bpf *, sysctl_monitor_bpf_free);
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct ring_buffer *, rb_free);
+
+static int sysct_write_event(void *ctx, void *data, size_t data_sz) {
+        struct sysctl_write_event *we = ASSERT_PTR(data);
+        Hashmap **sysctl_shadow = ASSERT_PTR(ctx);
+        _cleanup_free_ char *path = NULL;
+        char *value;
+
+        if (we->version != 1)
+                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                "Invalid sysctl write event version: %d", we->version);
```

Comment:
maybe say "Unexpected" instead of "Invalid".

what happens if you return negative here? does libbpf stop calling back?

i think it would be safe if we turn off the event source here, so that we aren't called back anymore if we hit this.

Also, say this in the error message "Unexpected sysctl event, disabling sysctl monitoring: %d"

Reply:
Unfortunately this is not documented in the docs so I readed the libbpf source code and it seems that returning a negative value stops the loop:
https://elixir.bootlin.com/linux/v6.10.9/source/tools/lib/bpf/ringbuf.c#L260

This is confirmed by this blog post:
https://nakryiko.com/posts/bpf-ringbuf/#:~:text=First%2C%20the%20definition,consumer%20loop

So I will return a negative value there

Reply:
and please add a comment that clarifies this to the casual reader of the sources.